src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb

```diff
@@ -1195,6 +1195,121 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+        ' if one of its instance fields is managed or a ref field.  Unfortunately, this can result in infinite recursion.
+        ' If the closure is finite, and we don't find anything definitely managed, then we return true.
+        ' If the closure is infinite, we disregard all but a representative of any expanding cycle.
+        ' 
+        ' Intuitively, this will only return true if there's a specific type we can point to that is would
+        ' be managed even if it had no fields.  e.g. struct S { S s; } is not managed, but struct S { S s; object o; }
+        ' is because we can point to object.
+        ' </summary>
+        Private Overloads Shared Function GetManagedKind(type As NamedTypeSymbol, ByRef useSiteInfo As CompoundUseSiteInfo(Of AssemblySymbol)) As ManagedKind
+            type = DirectCast(type.GetTupleUnderlyingTypeOrSelf(), NamedTypeSymbol)
+
+            ' The code below is a clone of BaseTypeAnalysis.GetManagedKind from C# it should be kept in sync going forward
+            Dim managedInfo = INamedTypeSymbolInternal.Helpers.IsManagedTypeHelper(type)
+            Dim definitelyManaged = (managedInfo.isManaged = ThreeState.True)
+
+            If managedInfo.isManaged = ThreeState.Unknown Then
+                ' Otherwise, we have to build and inspect the closure of depended-upon types.
+                Dim hs = PooledHashSet(Of Symbol).GetInstance()
+                Dim result = DependsOnDefinitelyManagedType(type, hs, useSiteInfo)
+                definitelyManaged = result.definitelyManaged
+                managedInfo.hasGenerics = managedInfo.hasGenerics OrElse result.hasGenerics
+                hs.Free()
+            End If
+
+            If definitelyManaged Then
+                Return ManagedKind.Managed
+            ElseIf managedInfo.hasGenerics Then
+                Return ManagedKind.UnmanagedWithGenerics
+            Else
+                Return ManagedKind.Unmanaged
+            End If
+        End Function
+
+        Private Shared Function DependsOnDefinitelyManagedType(
+            type As NamedTypeSymbol,
+            partialClosure As HashSet(Of Symbol),
+            ByRef useSiteInfo As CompoundUseSiteInfo(Of AssemblySymbol)
+        ) As (definitelyManaged As Boolean, hasGenerics As Boolean)
+
+            Debug.Assert(type IsNot Nothing)
+
+            Dim hasGenerics = False
+
+            If partialClosure.Add(type) Then
+
+                For Each member In type.GetMembersUnordered()
+
+                    Dim field = TryCast(member, FieldSymbol)
+
+                    ' Only instance fields (including field-like events) affect the outcome.
+                    If field Is Nothing OrElse field.IsShared Then
+                        Continue For
+                    End If
+
+                    useSiteInfo.Add(field.GetUseSiteInfo())
+                    Dim fieldType As TypeSymbol = field.Type.GetTupleUnderlyingTypeOrSelf()
```

Comment:
>GetTupleUnderlyingTypeOrSelf() [](http://example.com/codeflow?start=61&length=30)

Are we testing this case?

Reply:
> Are we testing a case where this call makes a difference?

I cannot think of a situation when it would make an observable difference in behavior. That is not to say that that it is impossible to run into one in some edge error scenario. Given the difference in the way C# and VB compilers represent tuple symbols, I prefer to be conservative and explicitly unwrap tuple types.